### ISSUE TITLE
Difference in training over single node and distributed training

### ISSUE THREAD

**Issue by abhimohta** [Microsoft/CNTK#2721]
I'm trying to do training leveraging multiple GPU nodes on my system, but the results that I get using a single node and using multiple nodes is different - rather very different. 

My code for training is as under -

```
for seq1, seq2, gt_scores in next_batch(train_data[0], train_data[1], train_data[2]):
            if(minibatch_count % C.Communicator.num_workers()) == C.Communicator.rank():
                trainer.train_minibatch({inp_1:seq1, inp_2:seq2, golden:gt_scores})
```
Any possible issues/errors on my end?

**Comment by eldakms** [comment]
Could you please also post how you created the trainer? Which distributed learner did you use?

**Comment by abhimohta** [comment]
I have exactly followed points 2 and 4 in this document: https://docs.microsoft.com/en-us/cognitive-toolkit/multiple-gpus-and-machines

**Comment by eldakms** [comment]
So which distributed learner are you using?

**Comment by abhimohta** [comment]
Adam is the learner that we are using. Please find the snippet of code below where we define the distributed learner -
```

    optimizer = C.adam(prediction.parameters,
                       lr=C.learning_rate_schedule(args.learning_rate, C.UnitType.minibatch),
                       momentum=C.momentum_schedule(0.9))
    distributed_after = args.batch_size
    distributed_learner = distributed.data_parallel_distributed_learner(learner = optimizer, num_quantization_bits = 32, distributed_after = 0)
    trainer = C.Trainer(prediction, (loss, error), distributed_learner, progress_writers=progress)

```

**Comment by eldakms** [comment]
I think there is a discrepancy in your minibatch size for distributed and non distributed case.
Built-in readers actually split a single minibatch of size N into two, then run on half of the minibatch locally, then aggregate gradients for the whole minibatch.

In your case you actually double the size of the minibatch by giving the full local minibatch to the trainer. So the aggregation happens for 2N samples instead of N.

Could you half it and try out? Thanks!
